design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,582 @@
+BareMetalHost visible to the end-user.
+
+All HardwareData are labeled with an ``infra-kind`` label whose
+value depends on the characteristics of the computer.
+
+* I can read HardwareData in namespace ``infra``. I can find which selector
+  I should use to find a machine with the right hardware details.
+* I create a resource with the following content:
+
+  ```yaml
+  apiVersion: metal3.io/v1alpha1
+  kind: HostClaim
+  metadata:
+    name: my-host
+  spec:
+    online: false
+
+    hostSelector:
+      matchLabels:
+        infra-kind: medium
+  ```
+
+* After a while, the system associates the claim with a real server, and
+  the resource's status is populated with the following information:
+
+  ```yaml
+  status:
+    bareMetalHost:
+      name: server-123
+      namespace: infra
+    conditions:
+    - lastTransitionTime: "2024-03-29T14:33:19Z"
+      status: "True"
+      type: Ready
+    - lastTransitionTime: "2024-03-29T14:33:19Z"
+      status: "True"
+      type: AssociateBMH
+    lastUpdated: "2024-03-29T14:33:19Z"
+  ```
+
+* I create three secrets in the same namespace ``my-user-data``,
+  ``my-metadata``, and ``my-network-data``. I use the information from the
+  status and meta data to customize the scripts they contain.
+* I modify the HostClaim to point to those secrets and start the server:
+
+  ```yaml
+  apiVersion: metal3.io/v1alpha1
+  kind: HostClaim
+  metadata:
+    name: my-host
+  spec:
+    online: true
+    image:
+      checksum: https://url_image.qcow2.md5
+      url: https://url_image.qcow2
+      format: qcow2
+    userData:
+      name: my-user-data
+    networkData:
+      name: my-network-data
+    hostSelector:
+      matchLabels:
+        infra-kind: medium
+  ```
+
+* The workload is launched. When the machine is fully provisioned, the boolean
+  field ready in the status becomes true. I can stop the server by changing
+  the online status. I can also perform a reboot by targeting specific
+  annotations in the reserved ``host.metal3.io`` domain.
+* When I destroy the host, the association is broken and another user can take
+  over the server.
+
+##### Server Administrator Point of View
+
+As the owner of the bare-metal servers, I would like to easily control the users
+that can use my servers and control the information I share about the servers.
+
+In each namespace hosting BareMetalHosts, I define one
+or several HostDeployPolicy resources. Each resource defines a filter on the
+namespaces of the HostClaims in the field ``hostclaimNamespaces``.
+
+The specification of the HostDeployPolicy contains one or several of the
+optional fields:
+
+* ``namespaces``: the content of this field is a list of namespaces. The name of
+  the namespace where the HostClaim is must be a member of the list.
+* ``nameMatches``: the content of this field is a regular expression
+  (in Golang re2 syntax). The name of the namespace where the HostClaim is located
+  must match this regular expression.
```

Comment:
nit: it looks like you could just use a single `namespaceMatches` list where each item is a regular expression

Reply:
> It is easy to add a new item to the list with a patch. 

Fair point.

> I am not sure we really need nameMatches

Yeah, maybe omit it from the initial version for simplicity.